docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+
+---
+### Provisioning
+```go
+// MachinePhaseProvisioning is the state when the
+// Machine infrastructure is being created.
+MachinePhaseProvisioning = MachinePhaseType("provisioning")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Pending”
+- `Machine.Spec.Bootstrap.ConfigRef`->`Status.Phase` is “Ready”
+- `Machine.Spec.Bootstrap.Data` is not `<nil>`
+
+#### Expectations
+- Machine’s infrastructure to be in the process of being provisioned.
+
+---
+### Provisioned
+```go
+// MachinePhaseProvisioned is the state when its
+// infrastructure has been created and configured.
+MachinePhaseProvisioned = MachinePhaseType("provisioned")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioning”
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Ready”
+
+#### Expectations
+- Machine’s infrastructure has been created and the compute resource is available to be configured.
+
+---
+### Running
+```go
+// MachinePhaseRunning is the Machine state when it has
+// become a Kubernetes Node in a Ready state.
+MachinePhaseRunning = MachinePhaseType("running")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioned”
+- `Machine.Spec.ProviderID` is not `<nil>`
+- A Kubernetes Node is found with the same `Machine.Spec.ProviderID` and its state is “Ready”
+
+#### Expectations
+- Machine controller should set `Machine.Status.NodeRef`.
+
+---
+### Deleting
+```go
+// MachinePhaseDeleting is the Machine state when a delete
+// request has been sent to the API Server,
+// but its infrastructure has not yet been fully deleted.
+MachinePhaseDeleting = MachinePhaseType("deleting")
+```
+
+#### Transition Conditions
+- `Machine.Phase` can be any value
+- `Machine.ObjectMeta.DeletionTimestamp` is not `<nil>`
+
+#### Expectations
+- Node associated should be drained and cordoned.
+- Machine’s related resources should be deleted first using cascading deletion.
+
+---
+### Deleted
+```go
+// MachinePhaseDeleted is the Machine state when the object
+// and the related infrastructure is deleted and
+// ready to be garbage collected by the API Server.
+MachinePhaseDeleted = MachinePhaseType("deleted")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Deleting”
+- `Machine.ObjectMeta.DeletionTimestamp` is not `<nil>`
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Deleted”
+
+#### Expectations
+- Machine controller should remove finalizer.
+
+---
+### Failed
+```go
+// MachinePhaseFailed is the Machine state when the system
+// might require user intervention.
+MachinePhaseFailed = MachinePhaseType("failed")
+```
+
+#### Transition Conditions
+- `Machine.ErrorReason` and/or `Machine.ErrorMessage` is populated.
+
+#### Expectations
+- User intervention.
+
+
+![Figure 1](https://i.imgur.com/VApZ8vv.png)
```

Comment:
nit: In the diagram, I think "post-deleting" is more appropriate than "post-deletion" as the name of the transition between the "deleting" and "deleted" states

Reply:
Fixed, thanks!